### DISCUSSION TITLE
How to assign two slices containing different types?

### DISCUSSION THREAD

**Discussion by listepo** [switchupcb/copygen#38] [Q&A] [answered]
For example:
```
type Copygen interface {
	ToCurrenciesContract(m []*models.Currency) []*contracts.Currency
}
```
generates:
```
// ToCurrenciesContract copies a []*models.Currency to a []*contracts.Currency.
func ToCurrenciesContract(tc []*contracts.Currency, fm []*models.Currency) {
	// []*contracts.Currency fields
}
```

**Comment by switchupcb** [comment, marked as answer]
You want to generate a function that converts all from objects `[]*models.Currency` to `[]*contracts.Currency`.
- `[]*models.Currency` is a slice of struct pointers to a `models.Currency` struct. 
- `[]*contracts.Currency` is a slice of struct pointers to a `contracts.Currency` struct.
- These types are not equivalent and won't be assigned to each other by default.

There are multiple interpretations of the request to assign a slice of one type to a slice of another type.
1. For each item in each slice, assign the respective fields from `contracts.Currency` to `models.Currency`.
2. For each item in each slice, assign the respective interface `contracts.Currency` to `models.Currency` object _(using type conversion)_.
3. For each item in each slice, assign the respective object `contracts.Currency` to interface `models.Currency`.
4. Any variation thereof...

Implementing a solution to any interpretation requires use of type conversion or assertion.
 
# Implementation

## v0.4

Use the `convert` option to implement this functionality.

### convert

```go
type Copygen interface {
	ToCurrenciesContract(m []*models.Currency) []*contracts.Currency
}

/* Define the function and field this converter is applied to using regex. */
// convert .* \[\]\*models.Currency
// ModelCurrenciesToContractCurrencies converts a slice of model currencies to a slice of contract currencies.
func ModelCurrenciesToContractCurrencies(m []*models.Currency) []*contracts.Currency{
        var output []*contracts.Currency

        // Your implementation according to the interpretation above goes here...

	return output
}
```

_Need more information on the `convert` option? Check out the [convert](https://github.com/switchupcb/copygen#convert) demonstration in the main example._

This should generate the following code.

```go
// ToCurrenciesContract copies a []*models.Currency to a []*contracts.Currency.
func ToCurrenciesContract(tc []*contracts.Currency, fm []*models.Currency) {
	// []*contracts.Currency fields
       tc = ModelCurrenciesToContractCurrencies(fm)
}
```

_If not, please report this problem in the [issue tracker](https://github.com/switchupcb/copygen/issues)._

**Reply by listepo** [reply]
@switchupcb thank you a lot

**Reply by switchupcb** [reply]
@listepo I have updated my example to replace the following lines.

```go
// convert .* \[\]\*models.Currency
```

This is because the first field of a `convert` option specifies which functions the conversion function is used in, and the second specifies which fields the function will be used on.
